### DISCUSSION TITLE
How to best handle immediatly-used class-scoped methods?

### DISCUSSION THREAD

**Discussion by Avasam** [python/typing#1624] [General]
Please see this file: https://github.com/pypa/setuptools/pull/4099/files
Looking for guidance as to why the existing form is disallowed or discouraged. Is there a better way to handle this scenario than I did?

> I find it annoying that code that was previously perfectly-valid Python is no longer allowed. I see now that these functions were used locally-only and ignored for their presence in the class. I wonder if the Python language has an opinion on this usage. That is, I'd really like to see an upstream reference (ideally in Python and not just mypy) that clarifies why this usage is discouraged and acknowledges/justifies the weakness(es):
> 
> > Functions that are used only within the scope of the class declaration and are otherwise unused/ignored after the class is constructed should be declared outside the class. Although this change violates principles of encapsulation, it's preferable to have these declared outside the class scope to avoid implying that such functions are members of the class.
> 
> On the other hand, it does appear as if staticmethod is the default condition:
> 
> ```
> >>> class Foo:
> ...   def bar(x):
> ...     print(x)
> ... 
> >>> Foo.bar('foo')
> foo
> ```
> 
> It really feels to me like mypy is broken here, and unless there's upstream guidance as to why the existing form is disallowed or discouraged, let's just suppress the mypy errors with `typing: ignore`.

\- Jaraco https://github.com/pypa/setuptools/pull/4099#issuecomment-1908580376

---

Leaving it as-is leads to 
Pyright: `Instance methods should take a "self" parameter [reportSelfClsParameterName]`
Pyright: `Argument of type "Self@AbstractSandbox" cannot be assigned to parameter "__name" of type "str" in function "getattr". "AbstractSandbox*" is incompatible with "str" [reportArgumentType]`
Mypy:
`No overload variant of "getattr" matches argument types Module, "AbstractSandbox"`

Specifying the type for `name: str`:
Pyright: `Instance methods should take a "self" parameter [reportSelfClsParameterName]`
Mypy: `Self argument missing for a non-static method (or an invalid type for self) [misc]`

Changing the method to a `classmethod` (with params `cls, name: str` doesn't work because I can't call `AbstractSandbox._mk_dual_path_wrapper` at the class-level.

Using a `staticmethod` leads to no type error (false-negative?) and the following error at runtime:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "E:\Users\Avasam\Documents\Git\setuptools\setuptools\sandbox.py", line 266, in <module>
    class AbstractSandbox:
  File "E:\Users\Avasam\Documents\Git\setuptools\setuptools\sandbox.py", line 314, in AbstractSandbox
    locals()[name] = _mk_dual_path_wrapper(name)
TypeError: 'staticmethod' object is not callable
```

**Comment by Daverball** [comment]
This just seems like a bad idea, technically type checkers could special-case calling functions that were defined in the class scope within the class scope, but there's already a ton of special casing to make methods, `classmethod` and `staticmethod` work and this is such a fringe use-case that I honestly don't see it come up often enough for type checkers to support it. So moving these utility functions outside the class body seems like the correct call.

Also the `staticmethod` error is appropriate, since this descriptor only became callable in Python 3.10, so in earlier versions you would not be able to call the descriptor itself. You can do `_mk_dual_path_wrapper.__func__(name)` which is backwards compatible.
